File: src/nostr/rpc_machine_test.rs
use super::*;

    fn mock_event(id: &str, pubkey: &str) -> Event {
        Event {
            id: id.to_string(),
            pubkey: pubkey.to_string(),
            created_at: 0,
            kind: 23194,
            tags: vec![],
            content: "".to_string(),
            sig: "".to_string(),
        }
    }

    #[test]
    fn test_rpc_machine_flow() {
        let req = mock_event("req1", "pk1");
        let mut machine = NwcRpcMachine::new(req);

        let actions = machine.start();
        assert_eq!(actions.len(), 2);
        assert!(matches!(actions[0], RpcAction::Subscribe(_, _)));
        assert!(matches!(actions[1], RpcAction::Publish(_)));
        assert_eq!(machine.state, RpcState::AwaitingResponse);

        // Feed EOSE - should be ignored
        let action = machine.transition(RelayMessage::Eose("rpc_sub".into()));
        assert!(action.is_none());
        assert_eq!(machine.state, RpcState::AwaitingResponse);

        // Feed Response EVENT
        let mut resp = mock_event("resp1", "pk2");
        resp.tags = vec![crate::nostr::Tag::E("req1".to_string(), vec![])];
        let action = machine.transition(RelayMessage::Event("rpc_sub".into(), resp.clone()));

        assert_eq!(action, Some(RpcAction::Unsubscribe("rpc_sub".into())));
        assert_eq!(machine.state, RpcState::Success(resp));
    }

    #[test]
    fn test_rpc_machine_unmatched_response() {
        let req = mock_event("req1", "pk1");
        let mut machine = NwcRpcMachine::new(req);
        machine.start();

        // Response with wrong e-tag should not match
        let mut resp = mock_event("resp1", "pk2");
        resp.tags = vec![crate::nostr::Tag::E("wrong_req".to_string(), vec![])];
        let action = machine.transition(RelayMessage::Event("rpc_sub".into(), resp));
        assert!(action.is_none());
        assert_eq!(machine.state, RpcState::AwaitingResponse);
    }

    #[test]
    fn test_rpc_machine_transition_before_start() {
        let req = mock_event("req1", "pk1");
        let mut machine = NwcRpcMachine::new(req);

        // Machine is still in Initial state. Any transition should be ignored.
        let action = machine.transition(RelayMessage::Eose("rpc_sub".into()));
        assert!(action.is_none());
        assert_eq!(machine.state, RpcState::Initial);
    }

    #[test]
    fn test_rpc_machine_notice_transition() {
        let req = mock_event("req1", "pk1");
        let mut machine = NwcRpcMachine::new(req);
        machine.start();

        let action = machine.transition(RelayMessage::Notice("rate limited".into()));
        assert_eq!(action, Some(RpcAction::Unsubscribe("rpc_sub".into())));
        assert_eq!(machine.state, RpcState::Failed("Relay notice: rate limited".into()));
    }

File: src/nostr/rpc_machine.test.ts
import { describe, it, expect } from 'vitest';
import { NwcRpcMachine } from './rpc_machine.ts';
import type { Event } from './event.ts';
import type { RelayMessage } from './nip01.ts';
import { Tag } from './tag.ts';

function mockEvent(id: string, pubkey: string): Event {
  return {
    id,
    pubkey,
    createdAt: 0,
    kind: 23194,
    tags: [],
    content: '',
    sig: '',
  };
}

describe('NwcRpcMachine', () => {
  it('full flow: start → EOSE ignored → response matched → success', () => {
    const req = mockEvent('req1', 'pk1');
    const machine = new NwcRpcMachine(req);

    const actions = machine.start();
    expect(actions.length).toBe(2);
    expect(actions[0].kind).toBe('subscribe');
    expect(actions[1].kind).toBe('publish');
    expect(machine.getState()).toEqual({ kind: 'awaitingResponse' });

    // EOSE should be ignored
    const eoseMsg: RelayMessage = { type: 'EOSE', subscriptionId: 'rpc_sub' };
    const eoseAction = machine.transition(eoseMsg);
    expect(eoseAction).toBeNull();
    expect(machine.getState()).toEqual({ kind: 'awaitingResponse' });

    // Matching response event
    const resp = mockEvent('resp1', 'pk2');
    resp.tags = [Tag.e('req1')];
    const eventMsg: RelayMessage = { type: 'EVENT', subscriptionId: 'rpc_sub', event: resp };
    const eventAction = machine.transition(eventMsg);

    expect(eventAction).toEqual({ kind: 'unsubscribe', subId: 'rpc_sub' });
    expect(machine.getState()).toEqual({ kind: 'success', event: resp });
  });

  it('ignores response with wrong e-tag', () => {
    const req = mockEvent('req1', 'pk1');
    const machine = new NwcRpcMachine(req);
    machine.start();

    const resp = mockEvent('resp1', 'pk2');
    resp.tags = [Tag.e('wrong_req')];
    const msg: RelayMessage = { type: 'EVENT', subscriptionId: 'rpc_sub', event: resp };
    const action = machine.transition(msg);

    expect(action).toBeNull();
    expect(machine.getState()).toEqual({ kind: 'awaitingResponse' });
  });

  it('ignores transitions before start', () => {
    const req = mockEvent('req1', 'pk1');
    const machine = new NwcRpcMachine(req);

    const msg: RelayMessage = { type: 'EOSE', subscriptionId: 'rpc_sub' };
    const action = machine.transition(msg);

    expect(action).toBeNull();
    expect(machine.getState()).toEqual({ kind: 'initial' });
  });

  it('transitions to failed on NOTICE', () => {
    const req = mockEvent('req1', 'pk1');
    const machine = new NwcRpcMachine(req);
    machine.start();

    const msg: RelayMessage = { type: 'NOTICE', message: 'rate limited' };
    const action = machine.transition(msg);

    expect(action).toEqual({ kind: 'unsubscribe', subId: 'rpc_sub' });
    expect(machine.getState()).toEqual({ kind: 'failed', reason: 'Relay notice: rate limited' });
  });
});
